bin/propolis-server/src/lib/vm/mod.rs

```diff
@@ -2,1181 +2,579 @@
-        // thread, which owns the sender side of the controller's state-change
-        // notification channel. Since the controller is being dropped, the
-        // worker is gone, so reclaim the sender from it and use it to publish
-        // that the controller is being destroyed.
-        if let Some(thread) = self.worker_thread.lock().unwrap().take() {
-            let api_state = thread.join().unwrap();
-            let old_state = api_state.borrow().clone();
-
-            // Preserve the instance's state if it failed so that clients can
-            // distinguish gracefully-stopped instances from failed instances.
-            if matches!(old_state.state, ApiInstanceState::Failed) {
-                return;
-            }
+    /// Moves this VM from the `Active` state to the `Rundown` state.
+    ///
+    /// This routine should only be called by the state driver.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the VM is not in the `Active` state.
+    async fn set_rundown(&self) {
+        info!(self.log, "setting VM rundown");
+        let services = {
+            let mut guard = self.inner.write().await;
+            let VmState::Active(vm) =
+                std::mem::replace(&mut guard.state, VmState::NoVm)
+            else {
+                panic!("VM should be active before being run down");
+            };
 
-            let gen = old_state.gen + 1;
-            let _ = api_state.send(ApiMonitoredState {
-                gen,
-                state: ApiInstanceState::Destroyed,
-                ..old_state
+            let spec = vm.objects().lock_shared().await.instance_spec().clone();
+            let ActiveVm { external_state_rx, properties, tokio_rt, .. } = vm;
+            guard.state = VmState::Rundown(VmDescription {
+                external_state_rx,
+                properties,
+                spec,
+                tokio_rt: Some(tokio_rt),
             });
-        }
+            vm.services
+        };
 
-        // Tokio will be upset if the VMM runtime is implicitly shutdown (via
-        // drop) in blocking context.  We avoid such troubles by doing an
-        // explicit background shutdown.
-        let rt = self.vmm_runtime.take().expect("vmm_runtime is populated");
-        rt.shutdown_background();
+        services.stop(&self.log).await;
     }
-}
 
-/// An event that a VM's state driver must process.
-#[derive(Debug)]
-enum StateDriverEvent {
-    /// An event that was raised from within the guest.
-    Guest(GuestEvent),
-
-    /// An event that was raised by an external entity (e.g. an API call to the
-    /// server).
-    External(ExternalRequest),
-}
-
-/// Commands issued by the state driver back to its VM controller. These are
-/// abstracted into a trait to allow them to be mocked out for testing without
-/// having to supply mock implementations of the rest of the VM controller's
-/// functionality.
-#[cfg_attr(test, mockall::automock)]
-trait StateDriverVmController {
```

Comment:
The functions in this trait moved to `VmObjects`, but the trait itself is gone--it existed to provide a way to mock these functions out for state driver testing. I haven't lost a lot of sleep over this because (a) PHD gives us a lot of signal as to whether these state transitions are working properly, and (b) mocks as test doubles in Rust have never sparked joy for me (compared to, say, the C++ codebase I worked in at $OLDJOB). I'm open to suggestions as to how to replace this (maybe some kind of `FakeVmObjects` is warranted, though I'm not sure exactly how it would look).